app/Main.hs

```diff
@@ -70,23 +72,29 @@ handleCliInput input = do
   -- Connect to delegate server
   -- TODO: port change
   -- FIXUP: parse actual adress and other params
-  delegateNumber <- read . fromJust <$> lookupEnv "DELEGATE_NUMBER"
+  delegateNumber :: Int <- do
+    let envVar = "DELEGATE_NUMBER"
+    lookupEnv envVar
+      >>= maybe (unrecoverable (UserError (EnvVarMissing envVar))) pure . (readMaybe =<<)
+
   runClient "127.0.0.1" (8000 + delegateNumber) "/" $ \client -> do
     currentDelegateStateRef <- newIORef initialState
+    let tracer = contramap (show . pretty) stdoutTracer
     sendTextData client . encode $ QueryCurrentDelegateState
     -- DelegateState receiving thread
-    withAsync (updateStateThread client currentDelegateStateRef) $ \_ -> do
-      handleCliInput' client currentDelegateStateRef input
+    withAsync (updateStateThread client currentDelegateStateRef tracer) $ \_ -> do
+      handleQueryAction client currentDelegateStateRef input
   where
-    updateStateThread client currentDelegateStateRef = forever $ do
-      mMessage <- decode <$> receiveData client
+    updateStateThread client currentDelegateStateRef tracer = forever $ do
+      mMessage <- eitherDecode <$> receiveData client
       case mMessage of
-        Just (CurrentDelegateState _ state) ->
+        Right (CurrentDelegateState _ state) ->
           writeIORef currentDelegateStateRef state
-        _ -> putStrLn $ "Ignoring server message: " <> show mMessage
+        Right response -> traceWith tracer (UnimplementedResponse response)
+        Left err -> traceWith tracer (InvalidResponse err)
 
-handleCliInput' :: _ -> IORef _ -> CliInput -> IO ()
-handleCliInput' client currentDelegateStateRef input = case input of
+handleQueryAction :: Connection -> IORef DelegateState -> CliInput -> IO ()
```

Comment:
Why is it `QueryAction`?

Reply:
I will change it back, I just changed it because it was easier to understand for me than `f`, `f'`, `f''`, ...